internal/monitors/statsd/statsd.go

```diff
@@ -0,0 +1,160 @@
+package statsd
+
+import (
+	"net"
+	"strconv"
+	"strings"
+)
+
+type statsDListener struct {
+	ipAddr       string
+	port         uint16
+	tcp          bool
+	udpConn      *net.UDPConn
+	tcpConn      *net.TCPConn
+	prefix       string
+	converters   []*converter
+	metricBuffer []string
+}
+
+type statsDMetric struct {
+	metricName string
+	metricType string
+	value      float64
+	dimensions map[string]string
+}
+
+func (sl *statsDListener) Listen() error {
+	if sl.tcp {
+		return sl.listenTCP()
+	}
+
+	return sl.listenUDP()
+}
+
+func (sl *statsDListener) listenUDP() error {
+	conn, err := net.ListenUDP("udp", &net.UDPAddr{
+		IP:   net.ParseIP(sl.ipAddr),
+		Port: int(sl.port),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	sl.udpConn = conn
+	return nil
+}
+
+func (sl *statsDListener) listenTCP() error {
+	return nil
+}
+
+func (sl *statsDListener) FetchMetrics() []*statsDMetric {
+	rawMetrics := make([]string, len(sl.metricBuffer))
+
+	copy(rawMetrics, sl.metricBuffer)
+	sl.metricBuffer = nil
+
+	parsed := parseMetrics(rawMetrics, sl.converters, sl.prefix)
+
+	return parsed
+}
+
+func (sl *statsDListener) Read() {
+	chData := make(chan []byte)
+
+	if sl.tcp {
+		go sl.readTCP(chData)
+	} else {
+		go sl.readUDP(chData)
+	}
+
+	for {
+		select {
+		case data := <-chData:
+			sl.metricBuffer = append(sl.metricBuffer, strings.Split(string(data), "\n")...)
+		}
+	}
+}
+
+func (sl *statsDListener) readTCP(chData chan []byte) {
+}
+
+func (sl *statsDListener) readUDP(chData chan []byte) {
+	// UDP needs to receive data packet by packet. Max packet size is 65535 for now.
+	buf := make([]byte, 65536)
+	for {
+		n, _, err := sl.udpConn.ReadFromUDP(buf)
+
+		if err != nil {
+			logger.WithError(err).Error("Failed reading UDP datagram.")
+			continue
+		}
+
+		chData <- buf[0:n]
+	}
+}
+
+func parseMetrics(raw []string, converters []*converter, prefix string) []*statsDMetric {
+	var metrics []*statsDMetric
+
+	for _, m := range raw {
+		colonIdx := strings.Index(m, ":")
+		pipeIdx := strings.Index(m, "|")
+		secondPipeIdx := pipeIdx + strings.Index(m[pipeIdx+1:], "|")
+
+		rawMetricName := m[0:colonIdx]
+		metricName := ""
+
+		var metricType string
+		if secondPipeIdx > pipeIdx {
+			metricType = m[pipeIdx+1 : secondPipeIdx]
+		} else {
+			metricType = m[pipeIdx+1:]
```

Comment:
What happens if the metric string ends in a `|`?  Just make sure it won't panic and crash the whole agent due to an out of bounds array index.